clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -24,123 +24,385 @@ package common
+
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) worker() {
+	for {
+		f, stop := w.queue.Get()
+		if stop {
+			return
+		}
+		(*f.(*func()))()
+		w.queue.Done(f)
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) gather(c clientset.Interface, syncTimeout time.Duration) error {
+	glog.Infof("%v: waiting for controlled pods measurement...", w)
+	if w.informer == nil {
+		return fmt.Errorf("metric %s has not been started", w)
+	}
+	desiredCount, maxResourceVersion, err := w.getObjectCountAndMaxVersion(c)
 	if err != nil {
-		return summaries, err
+		return err
+	}
+
+	if err := wait.Poll(checkControlledPodsInterval, syncTimeout, func() (bool, error) { return w.opResourceVersion >= maxResourceVersion, nil }); err != nil {
+		return fmt.Errorf("timed out while waiting for controlled pods")
+	}
+
+	w.handlingGroup.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	numberRunning := 0
+	for _, checker := range w.checkerMap {
+		if status := checker.getStatus(); status {
+			numberRunning++
+		}
+	}
+	if desiredCount != numberRunning {
+		glog.Errorf("%s: incorrect objects number: %d/%d %ss are running with all pods", w, numberRunning, desiredCount, w.kind)
+		return fmt.Errorf("incorrect objects number")
+	}
+
+	glog.Infof("%s: %d/%d %ss are running with all pods", w, numberRunning, desiredCount, w.kind)
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) stop() {
+	close(w.stopCh)
+	w.queue.ShutDown()
+	w.workerGroup.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.isRunning = false
+	for _, checker := range w.checkerMap {
+		checker.terminate()
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) handleObject(c clientset.Interface, oldObj, newObj interface{}) {
+	var oldRuntimeObj runtime.Object
+	var newRuntimeObj runtime.Object
+	var ok bool
+	oldRuntimeObj, ok = oldObj.(runtime.Object)
+	if oldObj != nil && !ok {
+		return
+	}
+	defer func() {
+		// We want to update version after (potentially) creating goroutine.
+		if err := w.updateOpResourceVersion(oldRuntimeObj); err != nil {
+			glog.Errorf("%s: updating resource version error: %v", w, err)
+		}
+	}()
+	newRuntimeObj, ok = newObj.(runtime.Object)
+	if newRuntimeObj != nil && !ok {
+		return
+	}
+	defer func() {
+		// We want to update version after (potentially) creating goroutine.
+		if err := w.updateOpResourceVersion(newRuntimeObj); err != nil {
+			glog.Errorf("%s: updating resource version error: %v", w, err)
+		}
+	}()
+	if isEqual, err := runtimeobjects.IsEqualRuntimeObjectsSpec(oldRuntimeObj, newRuntimeObj); err != nil || isEqual {
+		if err != nil {
+			glog.Errorf("%s: comparing specs error: %v", w, err)
+		}
+		return
 	}
-	labelSelector, err := util.GetStringOrDefault(config.Params, "labelSelector", "")
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	if !w.isRunning {
+		return
+	}
+
+	if oldRuntimeObj != nil {
+		if err := w.deleteObjectLocked(c, oldRuntimeObj); err != nil {
+			glog.Errorf("%s: delete checker error: %v", w, err)
```

Comment:
I just realized that a bunch of those glog.Errors shoudl actually fail the test (in the end, not immediately).
glog.Error doesn't have that property.

Reply:
That's true. However, I expect that all small errors will have impact on final measurement check (e.g. failing to check object deletion will result on failing object count check), so all in all whole measurement will fail.